### ISSUE TITLE
How to make ScrollTransitions as in SwiftUI in Flutter

### ISSUE THREAD

**Issue by letsar** [flutter/flutter#128870]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Context

SwiftUI introduced a great way to make beautiful scroll transitions easily as in the video below:

https://github.com/flutter/flutter/assets/9378033/274e44d0-6dc1-4f67-8158-0a6cc9486b7e

This video comes from this tweet: https://twitter.com/i/status/1666854635796148244

We can find the API documentation [here](https://developer.apple.com/documentation/swiftui/view/scrolltransition(topleading:bottomtrailing:axis:transition:)).

The code simply looks like this (this is not the code for the video above, it's just to display some code).
```swift
CellView()
    .idView(i)
    .scrollTransition(.animated) { content, phase in
        content
            .scaleEffect(phase != .identity ? 0.6 : 1)
            .opacity(phase != .identity ? 0.3 : 1)
    }
```

I don't this think we can do this as easily in Flutter today.

### Objectives
- Add the API primitives to let a child get rendering information from one of its parents.
- It should be done in the same frame the child is rendered

### First ideas
I wanted to see how to tackle this.

First I thought it would be possible to pass a different constraint type when calling the `layout` method of the `child` in the appropriate `RenderObject`. Then with a specific `LayoutBuilder` it would have been possible to get the scroll information.
But it can't work because we need to have the size of the child in order to compute the visible portion of the child within the scrollview.

So because we need the size of the child, we can't do it before the painting phase.
I looked at how we could pass information through the `PaintingContext` class, I thought about creating a `DelegatingPaintingContext` (which would proxy all the method to an inner `PaintingContext` instance) and add it some additional information concerning the scroll information.
Sadly it cannot be done outside the library where `PaintingContext` is. Moreover, I'm not sure the instance of the painting context is the same for all children in the tree. If a new instance (from `PaintingContext`) is created before the appropriate child get it, it won't work.

Another solution could be to do it at the compositing level by adding some layers, but I'm not sure it would be the best way to tackle this.

Maybe you could help me to choose the best direction to add this feature to Flutter. Through an external package if there is already a way to do it in Flutter I'm not aware of, or I can create a PR in Flutter if I have enough information to start working on this.

If I don't provided enough information, let me know and I'll update the description here.

### Proposal

For the moment I don't have a proposal.

**Comment by maRci002** [comment]
> But it can't work because we need to have the size of the child in order to compute the visible portion of the child within the scrollview.

Check out this cookbook, this will help implementeing your own effects
https://docs.flutter.dev/cookbook/effects/parallax-scrolling

**Comment by letsar** [comment]
@maRci002 it really look uneasy to do and not efficient to me since the child has to be rendered a first time before getting its `renderObject`.
Each time we need to access the `renderObject` through the context I feel like something is missing. To me widgets should not access their `renderObject`.
Also, to me GlobalKey should be avoided as much as possible.

**Comment by maRci002** [comment]
> @maRci002 it really look uneasy to do and not efficient to me since the child has to be rendered a first time before getting its `renderObject`. Each time we need to access the `renderObject` through the context I feel like something is missing. To me widgets should not access their `renderObject`. Also, to me GlobalKey should be avoided as much as possible.

I created a little demo that avoids `RenderBox` / `GlobalKey` and also uses `CustomScrollView`.

<details>
  <summary>Click me</summary>
  
  ```dart
  import 'package:flutter/material.dart';
  
  void main() {
    runApp(const MyApp());
  }
  
  class MyApp extends StatelessWidget {
    const MyApp({super.key});
  
    @override
    Widget build(BuildContext context) {
      return MaterialApp(
        theme: ThemeData.dark().copyWith(scaffoldBackgroundColor: const Color.fromARGB(255, 18, 32, 47)),
        debugShowCheckedModeBanner: false,
        home: const Scaffold(
          body: ExampleTransform(),
        ),
      );
    }
  }
  
  class ExampleTransform extends StatelessWidget {
    const ExampleTransform({super.key});
  
    @override
    Widget build(BuildContext context) {
      return CustomScrollView(
        primary: true,
        slivers: [
          SliverFixedExtentList(
            itemExtent: 345,
            delegate: SliverChildBuilderDelegate(
              childCount: locations.length,
              (context, index) {
                final location = locations[index];
  
                return _ApplyTransform(
                  itemOffset: index * 345,
                  child: LocationListItem(
                    imageUrl: location.imageUrl,
                    name: location.name,
                    country: location.place,
                  ),
                );
              },
            ),
          ),
        ],
      );
    }
  }
  
  class _ApplyTransform extends StatelessWidget {
    const _ApplyTransform({
      required this.itemOffset,
      required this.child,
    });
  
    final double itemOffset;
    final Widget child;
  
    @override
    Widget build(BuildContext context) {
      final controller = Scrollable.of(context);
  
      return AnimatedBuilder(
        animation: controller.widget.controller!,
        builder: _builder,
        child: child,
      );
    }
  
    Widget _builder(BuildContext context, Widget? child) {
      final controller = Scrollable.of(context);
      final position = controller.position;
  
      final offset = position.pixels;
      final viewPort = position.viewportDimension;
  
      final itemPosition = offset - itemOffset;
      final topBoundary = viewPort * 0.2;
  
      final itemPositionFraction = (itemPosition / topBoundary).clamp(0, 1);
  
      final rotationAngle = -0.1 * itemPositionFraction;
      final opacity = (1.0 - itemPositionFraction).clamp(0.3, 1.0);
  
      return Opacity(
        opacity: opacity,
        child: Transform(
          alignment: Alignment.center,
          transform: Matrix4.rotationZ(rotationAngle),
          child: child,
        ),
      );
    }
  }
  
  class LocationListItem extends StatelessWidget {
    const LocationListItem({
      super.key,
      required this.imageUrl,
      required this.name,
      required this.country,
    });
  
    final String imageUrl;
    final String name;
    final String country;
  
    @override
    Widget build(BuildContext context) {
      return Padding(
        padding: const EdgeInsets.symmetric(horizontal: 24, vertical: 16),
        child: AspectRatio(
          aspectRatio: 16 / 9,
          child: ClipRRect(
            borderRadius: BorderRadius.circular(16),
            child: _buildParallaxBackground(context),
          ),
        ),
      );
    }
  
    Widget _buildParallaxBackground(BuildContext context) {
      return Stack(
        children: [
          Image.network(
            imageUrl,
            fit: BoxFit.cover,
            width: double.infinity,
            height: double.infinity,
          ),
          _buildGradient(),
          _buildTitleAndSubtitle(),
        ],
      );
    }
  
    Widget _buildGradient() {
      return Positioned.fill(
        child: DecoratedBox(
          decoration: BoxDecoration(
            gradient: LinearGradient(
              colors: [Colors.transparent, Colors.black.withOpacity(0.7)],
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
              stops: const [0.6, 0.95],
            ),
          ),
        ),
      );
    }
  
    Widget _buildTitleAndSubtitle() {
      return Positioned(
        left: 20,
        bottom: 20,
        child: Column(
          mainAxisSize: MainAxisSize.min,
          crossAxisAlignment: CrossAxisAlignment.start,
          children: [
            Text(
              name,
              style: const TextStyle(
                color: Colors.white,
                fontSize: 20,
                fontWeight: FontWeight.bold,
              ),
            ),
            Text(
              country,
              style: const TextStyle(
                color: Colors.white,
                fontSize: 14,
              ),
            ),
          ],
        ),
      );
    }
  }
  
  class Location {
    const Location({
      required this.name,
      required this.place,
      required this.imageUrl,
    });
  
    final String name;
    final String place;
    final String imageUrl;
  }
  
  const urlPrefix = 'https://docs.flutter.dev/cookbook/img-files/effects/parallax';
  const locations = [
    Location(
      name: 'Mount Rushmore',
      place: 'U.S.A',
      imageUrl: '$urlPrefix/01-mount-rushmore.jpg',
    ),
    Location(
      name: 'Gardens By The Bay',
      place: 'Singapore',
      imageUrl: '$urlPrefix/02-singapore.jpg',
    ),
    Location(
      name: 'Machu Picchu',
      place: 'Peru',
      imageUrl: '$urlPrefix/03-machu-picchu.jpg',
    ),
    Location(
      name: 'Vitznau',
      place: 'Switzerland',
      imageUrl: '$urlPrefix/04-vitznau.jpg',
    ),
    Location(
      name: 'Bali',
      place: 'Indonesia',
      imageUrl: '$urlPrefix/05-bali.jpg',
    ),
    Location(
      name: 'Mexico City',
      place: 'Mexico',
      imageUrl: '$urlPrefix/06-mexico-city.jpg',
    ),
    Location(
      name: 'Cairo',
      place: 'Egypt',
      imageUrl: '$urlPrefix/07-cairo.jpg',
    ),
  ];
  
  ```
</details>


https://github.com/flutter/flutter/assets/8436039/5049ba08-a1f2-4655-9bd7-710581cc2e94

**Comment by letsar** [comment]
Thanks @maRci002, but I see two issues here which not satisfy the requirements I emitted:

In the sample a `SliverFixedExtentList` is used, which allow to know the size of each item before rendering. I would want to be able to do it without forcing the tight constraint on the children. Typically with a `SliverList`, we cannot know in advance the height of each child.

We use `Scrollable.of(context)` in this exemple to get the viewport dimensions and the position of the scroll. By using this we cannot fulfill the requirement saying that we want to do the transform on the same frame as the build. We would apply the real transform one frame after we render the child.

**Comment by letsar** [comment]
I modified `PaintingContext` to add a data field which would be used as a bag of rendering information:

```dart
class PaintingContext {
  ...
  Map<Type, Object>? _data;

  T? get<T>() {
    final Map<Type, Object>? data = _data;
    if (data == null) {
      return null;
    }
    return data[T] as T?;
  }

  void set<T extends Object>(T value) {
    final Map<Type, Object>? data = _data;
    if (data == null) {
      _data = <Type, Object>{
        T: value,
      };
    } else {
      data[T] = value;
    }
  }
  ...
}
```

It seems to work as I wanted but I don't know if there are any drawback to this idea.
@Piinks, @Hixie do you think this is something that is reasonable to add to Flutter? I can make a PR.

**Comment by letsar** [comment]
@maRci002 I may have been wrong about `Scrollable.of(context)` providing the real position one frame later. `ScrollViews` are built on `Scrollable` so it may not have any difference between getting the information into `RenderSliverMultiBoxAdaptor` or through `Scrollable.of(context)`!
If that's the case, I can build a package without touching to the Flutter SDK!

**Comment by Hixie** [comment]
This is an interesting problem.

I don't think passing around untyped data is very much in the style of the Flutter framework, so I probably wouldn't go for the PaintingContext suggestion.

You should be able to get the effect shown in the videos above using a custom RenderSliver, but that's definitely not as easy as what SwiftUI is showing.

I wonder if this is something https://pub.dev/packages/flutter_animate could add...

**Comment by letsar** [comment]
I tried to see how to add this as extensible as possible, but yes it would not be very flutterish.

My last idea would just be to add additional properties to SliverMultiBoxAdaptorParenData so that we can get them in a RenderBox child through an external package. But since it would probably not be used directly in Flutter, I don't know if you'll agree to add it.

A custom Sliver would not work I think, because otherwise we would have to set all items as slivers inside a CustomScrollView and thus we would lose the ability to build them lazily.
That's why I think it would be great if we could obtain the necessary data directly through the parentData of a parent.

I have a POC which works with only "small" changes into SliverMultiBoxAdaptorParenData. But if you see a better way, I'm all ears.

**Comment by letsar** [comment]
I think I may have found a way to do it from my side without any change to the Flutter SDK. Within a `RenderObject`, I can walk through its parents to get the parents of interest (the `SliverMultiBoxAdaptor` and its direct child), then I can get the sliver constraints and the child's parent data to get the information I want.

Is this something you already do in Flutter? I'm not very satisfied with this solution, but I don't really see a better way to do it at the moment.
Changing the Flutter SDK to modify the `SliverMultiBoxAdaptorParenData` would not change the way I get the parent data in the first place, so I might as well get the sliver constraints from the sliver.

@Hixie do you see any drawbacks, doing it like I plan to do?

If it's not clear, I can elaborate with some code.

**Comment by Piinks** [comment]
I wonder if instead this could be implemented as a transformation that is applied to the children of a sliverList based on the position.. Looking at the swift code, it chains transformations on the `items`. I don't think we have a SliverTransform, but if we want to support being able to apply transformations (of all sorts!) like based on the scroll position, I might start there. :)

**Comment by letsar** [comment]
@Piinks Yes it is indeed a set of paint transformations applied to the children of a `SliverList` based on their position.
I'm not sure we can use a `SliverTransform` here because the `SliverTransform` cannot be a child of a `SliverList` (only `RenderBox` children are accepted I think).
If we use a `SliverTransform`, then we would need to use a `CustomScrollView` and provide all the `SliverTransforms` as its children. By doing so, we could not lazily build them like it's the case with `SliverList`, and thus it would not work for a lot of cases.

**Comment by Piinks** [comment]
Yeah, totally, I was thinking about this too after I commented. I don't think we would want to have a SliverList that takes slivers 🙃 , that could get unnecessarily complex, or as you mentioned a bunch of non lazy slivers in a CustomScrollView. 

This makes me think of ReorderableList (or SliverReorderableList). It's  essentially a wrapper around SliverList for a specialized use case (dragging to reorder items). In this case, I might try a SliverLayoutBuilder that wraps a SliverList, so the provided SliverConstraints from the LayoutBuilder are available to transform the children provided to the SliverList. 🤔 

There are likely a bunch of better/other ways, these ideas are just a few. I think if we can come up with an abstract way to support child transformations on a SliverList, that sounds completely reasonable to me to include in the framework. :)

I'll file a separate bug for the SliverTransform, I think outside of this use case it may be worth investigating as well. We've been adding more slivers lately to match some of the box widgets that already exist.

**Comment by letsar** [comment]
Oh yes it would become too complicated if `SliverList` could take slivers as children 😅 !

Using `SliverLayoutBuilder` might be a could idea yes, but the issue then would be to create a specific version for all the existing slivers then (if we want to support it for all kind of slivers like `SliverGrid`, not only `SliverList`). I will think about this idea, it could lead to something.
But the other issue after that is to get the position of the direct children of the `Sliver`. Beside getting their `SliverMultiBoxAdaptorParenData`, I don't see how to do it.

I would be really great if we could come up with an abstract way to do this and include it in the framework 😃 ! In the meantime I can create a package with the _not so good_ solution I've found for people wanting to do it quickly, and I'll deprecate it when the abstract way would be in the framework.

Great for SliverTransform, it can be useful indeed :-).

**Comment by letsar** [comment]
@Piinks I've just release a version of a package to port the SwiftUI's API style to Flutter.
If you're interested you can find the code [here](https://github.com/letsar/visual_effect).
I've some issues to make work shaders and having a perfect start. For the moment with this solution, if the child's height is dynamic (as it can be with images), the package may not be able to get the right data after everything is stable. This is because I had to made my `RenderObject` a `RepaintBoundary` and repaint on scroll. I would prefer to be able to always paint when the `Sliver` parent repaints (by deactivating the automatic repaint boundaries though). I'm not enough experimented in that domain for the moment, so if you have an hint to make this work, let me know 🙂 .

**Comment by letsar** [comment]
Little update from my side. I succeeded to get it right at the first frame I think, so this experiment starts to take shape.
If someone has time to look at the code, I would appreciate the feedback to see if it's the right thing to do or it would be better to tackle this with another angle.
